pkg/build/webhook/generic/generic.go

```diff
@@ -71,19 +71,17 @@ func (p *WebHookPlugin) Extract(buildCfg *api.BuildConfig, secret, path string,
 			return nil, true, nil
 		}
 
-		if data.Git.Refs != nil {
-			for _, ref := range data.Git.Refs {
-				if webhook.GitRefMatches(ref.Ref, git.Ref) {
-					revision = &api.SourceRevision{
-						Git: &ref.GitSourceRevision,
-					}
-					return revision, true, nil
+		for _, ref := range data.Git.Refs {
+			if webhook.GitRefMatches(ref.Ref, &buildCfg.Spec.Source) {
+				revision = &api.SourceRevision{
+					Git: &ref.GitSourceRevision,
 				}
+				return revision, true, nil
 			}
-			glog.V(2).Infof("Skipping build for BuildConfig %s/%s. None of the supplied refs matched %q", buildCfg.Namespace, buildCfg, git.Ref)
+			glog.V(2).Infof("Skipping build for BuildConfig %s/%s. None of the supplied refs matched %q", buildCfg.Namespace, buildCfg, buildCfg.Spec.Source.Git.Ref)
 			return nil, false, nil
```

Comment:
also since you removed the check for nil, you're always going to return false without checking the data.Git.Ref field, so you need to put that nil check back in.

which again means we need better tests.